### PR TITLE
Add constant STATIC_INITIALIZER for allowBlockingCalls

### PR DESCRIPTION
**Pull request by simonbasle**
This commit introduces a STATIC_INITIALIZER constant in BlockHound, so
it becomes easier to discover that the features relying on the
`String methodName` parameter can work with `<clinit>` for static
initializer.

Additionally, this is documented in the customization examples and
used in the StaticInitTest.

Fixes #176.

### CODE REVIEW

**Comment by simonbasle** [comment]
(oops, the branch is misnamed and should refer to #178 rather than 176)

**Comment by simonbasle** [comment]
cc @rstoyanchev @bsideup

**Review by rstoyanchev — COMMENTED** [review]
I like the idea but could you please also add a paragraph in the Javadoc of allow and disallow that explicitly mentions this scenario?:
```
<p>To allow blocking in the static initializer block of a class, use the constant `{@link #STATIC_INITIALIZER} as the method name`.

**Comment by simonbasle** [comment]
> I like the idea but could you please also add a paragraph in the Javadoc of allow and disallow that explicitly mentions this scenario?

Done 👍

File: agent/src/main/java/reactor/blockhound/BlockHound.java [hunk 1]
```diff
@@ -62,7 +62,13 @@
  */
 public class BlockHound {
 
-    static final String PREFIX = "$$BlockHound$$_";
+    /**
+     * The special methodName that should be used in e.g. {@link Builder#allowBlockingCallsInside(String, String)}
+     * when one wants to reference the static initializer block rather than a class method.
+     */
+    public static final String STATIC_INITIALIZER = "<clinit>";
```
Comment (bsideup) [inline agent/src/main/java/reactor/blockhound/BlockHound.java:69]: since this constant would never change (part of the JVM's spec), I do not feel that it makes sense to introduce it. We could keep the docs section (for both `<clinit>` and `<init>`), but the constant will only create an unnecessary noise in the public signature of this class.

Comment (simonbasle) [reply agent/src/main/java/reactor/blockhound/BlockHound.java:69]: I don't know. Mentioning the reserved method name in the doc indeed feels like a minimum viable solution, but feels slightly less user-friendly than providing a safe constant which can quickly be auto-completed. Would it limit the noise drawback if the constant was defined in `Builder` instead of `BlockHound`? @royclarkson as a user, what do you think?

Comment (rstoyanchev) [reply agent/src/main/java/reactor/blockhound/BlockHound.java:69]: I also prefer an explicit API with documentation as a minimum. That said, thinking more about the constant it also gets slightly verbose with the class name included, and it's not automatically discoverable. 

What about adding a sub-builder step:
```java
builder
    .allowBlockingCallsInside("com.example.ClassA", "methodA")
    .allowBlockingCallsInside("com.example.ClassB").staticInitializers()
    .disallowBlockingCallsInside("com.example.ClassC").constructors();
```

The sub-builder something like:
```java
interface BlockingCallSpec {

    BlockHound.Builder staticInitializers();

    BlockHound.Builder constructors();

}
```

Or perhaps an enum after all?
```java
builder
    .allowBlockingCallsInside("com.example.ClassA", "methodA")
    .allowBlockingCallsInside("com.example.ClassB", BlockHound.TargetType.STATIC_INITIALIZER)
    .disallowBlockingCallsInside("com.example.ClassC", BlockHound.TargetType.CONSTRUCTORS);
```

**Comment by simonbasle** [comment]
@rstoyanchev @bsideup dc9186b introduces a `BuilderAllowSpec`. Since we only support static init for now, maybe it is a bit overkill. It would be even worse if we wanted to prevent going back to the `Builder` (via `and()`) without having configured at least one method first (as this would mean introducing 2 interfaces for the builder, so that `.allowBlockingCallsInside("foo").and()` isn't possible)

File: agent/src/main/java/reactor/blockhound/BlockHound.java [hunk 2]
```diff
@@ -254,6 +258,58 @@ public TypePool typePool(ClassFileLocator classFileLocator, ClassLoader classLoa
...
+             * class. Use {@link #and()} to exit the class-specific phase and return
+             * to the {@link Builder}.
+             *
+             * @return this {@link BuilderAllowSpec} for further configuration of the given class
+             * @see #and()
+             */
+            public BuilderAllowSpec forStaticInitializer() {
+                Builder.this.allowBlockingCallsInside(this.className, "<clinit>");
+                return this;
+            }
```
Comment (rstoyanchev) [inline agent/src/main/java/reactor/blockhound/BlockHound.java:288]: Why can't this go back to `Builder` and eliminate the need for `.and()`? Either way, as it is right now, the classname is repeated for every additional method. This would be no different. It would look like this:
```java
builder
    .allowBlockingCallsInside("...").forStaticInitializer()
    .allowBlockingCallsInside("...").forMethod("...")
   ...

```

File: agent/src/main/java/reactor/blockhound/BlockHound.java [hunk 3]
```diff
@@ -254,6 +258,58 @@ public TypePool typePool(ClassFileLocator classFileLocator, ClassLoader classLoa
...
+             * class. Use {@link #and()} to exit the class-specific phase and return
+             * to the {@link Builder}.
+             *
+             * @return this {@link BuilderAllowSpec} for further configuration of the given class
+             * @see #and()
+             */
+            public BuilderAllowSpec forMethod(String methodName) {
+                Builder.this.allowBlockingCallsInside(this.className, methodName);
+                return this;
+            }
```
Comment (rstoyanchev) [inline agent/src/main/java/reactor/blockhound/BlockHound.java:301]: Couldn't this be a vararg to allow multiple method names?

File: agent/src/main/java/reactor/blockhound/BlockHound.java [hunk 4]
```diff
@@ -254,6 +258,58 @@ public TypePool typePool(ClassFileLocator classFileLocator, ClassLoader classLoa
...
+
+            /**
+             * Exit the class-specific phase and return to the {@link Builder}.
+             *
+             * @return the {@link Builder}
+             */
+            public Builder and() {
+                return Builder.this;
+            }
+        }
```
Comment (rstoyanchev) [inline agent/src/main/java/reactor/blockhound/BlockHound.java:311]: What about constructors?

File: agent/src/main/java/reactor/blockhound/BlockHound.java [hunk 5]
```diff
@@ -254,6 +258,58 @@ public TypePool typePool(ClassFileLocator classFileLocator, ClassLoader classLoa
...
 
+        /**
+         * A phase of the builder where allowed blocking methods are configured for a given class.
+         * See {@link Builder#allowBlockingCallsInside(String)}.
+         * <p>
+         * Note that methods of this class directly mutate the {@link Builder} as if one was calling
+         * {@link Builder#allowBlockingCallsInside(String, String)}. One MUST specify at least one
+         * allowed method for it to have an effect on the configuration.
+         */
+        public class BuilderAllowSpec {
```
Comment (rstoyanchev) [inline agent/src/main/java/reactor/blockhound/BlockHound.java:269]: I don't see anything for "Disallow". Doesn't that need the same? The same builder could support both.

**Review by rstoyanchev — CHANGES_REQUESTED** [review]

File: agent/src/main/java/reactor/blockhound/BlockHound.java (hunk 1 above)
Comment (simonbasle) [reply agent/src/main/java/reactor/blockhound/BlockHound.java]: exploring the builder idea in this PR and the enum in https://github.com/reactor/BlockHound/pull/184

File: agent/src/main/java/reactor/blockhound/BlockHound.java (hunk 4 above)
Comment (simonbasle) [reply agent/src/main/java/reactor/blockhound/BlockHound.java:311]: ByteBuddy currently doesn't seem to support constructors, at least not with the way the agent is set up to apply allowances

Comment (rstoyanchev) [reply agent/src/main/java/reactor/blockhound/BlockHound.java:311]: If there is nothing that can be done, this should at least be documented as a limitation to save the time if digging in, when the need arises (as it did in my case) only to find out it's not possible.

File: agent/src/main/java/reactor/blockhound/BlockHound.java (hunk 2 above)
Comment (simonbasle) [reply agent/src/main/java/reactor/blockhound/BlockHound.java]: good point, done

File: agent/src/main/java/reactor/blockhound/BlockHound.java (hunk 3 above)
Comment (simonbasle) [reply agent/src/main/java/reactor/blockhound/BlockHound.java:280]: done (renamed `forMethods`. has the added benefit that there is actually a difference between the original `allowBlockingCallsInside(String, String)` and this one => it can handle multiple methods in one call)

File: agent/src/main/java/reactor/blockhound/BlockHound.java (hunk 5 above)
Comment (simonbasle) [reply agent/src/main/java/reactor/blockhound/BlockHound.java:264]: it could, but I wasn't able to test the disallow feature. Either I misunderstand it or there is something wrong with the feature (which isn't tested)... so for now I prefer to focus on the allow-static-initializer case which is demonstrably supported.

File: agent/src/main/java/reactor/blockhound/BlockHound.java (hunk 4 above)
Comment (simonbasle) [reply agent/src/main/java/reactor/blockhound/BlockHound.java:295]: would mentioning the limitation (and the #174 issue) in the root javadoc of `BuilderAllowSpec` work?

File: agent/src/main/java/reactor/blockhound/BlockHound.java (hunk 5 above)
Comment (bsideup) [reply agent/src/main/java/reactor/blockhound/BlockHound.java:264]: @simonbasle re "disallow": see https://github.com/reactor/BlockHound/blob/master/docs/customization.md#dis-allowing-blocking-calls-inside-methods

**Review by bsideup — CHANGES_REQUESTED** [review]
I do not support the idea of adding DSL just for the sake of marking static constructors.
Also, disallowing should be supported.

Last but not least, it seems that static constructor is the only type of methods added by it, which only confirms that a dedicated DSL is an overkill

File: agent/src/main/java/reactor/blockhound/BlockHound.java (hunk 5 above)
Comment (simonbasle) [reply agent/src/main/java/reactor/blockhound/BlockHound.java:267]: I saw that and didn't find it crystal clear.
The following tests fail:

```java
public class BlockingDisallowTest {

    static {
        BlockHound.install(b -> b
                .disallowBlockingCallsInside("reactor.core.publisher.Flux", "just")
                .disallowBlockingCallsInside(NonBlockingClass.class.getName(), "example")
        );
    }

    @Test
    public void shouldDisallowFluxJust() {
        assertThatExceptionOfType(RuntimeException.class)
                .isThrownBy(() ->
                        Mono.just("one")
                            .publishOn(Schedulers.parallel())
                            .flatMapMany(i -> Flux.just("one", "two"))
                            .blockLast()
                )
                .withCauseInstanceOf(BlockingOperationError.class);

    }

    @Test
    public void shouldDisallowNonBlocking() {
        NonBlockingClass nbc = new NonBlockingClass();

        assertThatExceptionOfType(RuntimeException.class)
                .isThrownBy(() ->
                        Mono.fromCallable(nbc::example)
                            .subscribeOn(Schedulers.parallel())
                            .block()
                )
                .withCauseInstanceOf(BlockingOperationError.class);
    }

    static class NonBlockingClass {

        String example() {
            return "example";
        }
    }
}
```
What am I missing?

Comment (bsideup) [reply agent/src/main/java/reactor/blockhound/BlockHound.java:267]: You seem to be missing how BlockHound works, see https://github.com/reactor/BlockHound/blob/master/docs/how_it_works.md and especially "Blocking call decision".

tl;dr: `disallowBlockingCallsInside` disallows blocking calls inside provided method (as per the API method's name). It does not, however, mark the method as blocking, so your `Flux.just` or `NonBlockingClass.example` makes no sense.
